source/beamline/CLS/CLSAMDSScaler.h

```diff
@@ -0,0 +1,318 @@
+#ifndef CLSAMDSSCALER_H
+#define CLSAMDSSCALER_H
+
+#include "beamline/AMControlSet.h"
+
+class CLSAMDSScalerChannel;
+class AMAction3;
+class AMControl;
+class AMReadOnlyPVControl;
+class AMDetectorTriggerSource;
+class AMDetectorDwellTimeSource;
+class AMCurrentAmplifier;
+
+class AMDSClientRequest;
+class AMDSClientRelativeCountPlusCountDataRequest;
+
+#include "util/AMRange.h"
+#include "source/beamline/AMDetector.h"
+
+#define CLSAMDSScaler_NOT_CONNECTED_OR_IS_SCANNING 456000
+
+/*!
+  Builds an abstraction for the AMDS Interface to the scaler used throughout the CLS.  It takes in a base name of the PV's and builds all the PV's
+  and makes the necessary connections.
+  */
+class CLSAMDSScaler : public QObject
+{
+Q_OBJECT
+
+public:
+	/// Constructor.  Takes the baseName of the PV's as parameters.
+	CLSAMDSScaler(const QString &baseName, const QString &amdsBufferName = QString(), QObject *parent = 0);
+	/// Destructor.
+	virtual ~CLSAMDSScaler();
+
+	/// Returns whether the scaler is all connected.
+	bool isConnected() const;
+
+	/// Returns the buffer name for this detector
+	QString amdsBufferName() const;
+	/// Configures the server with the given identifier
+	void configAMDSServer(const QString &amdsServerIdentifier);
+	/// Fetches data if an AMDS Server is set up
+	bool retrieveBufferedData(double seconds);
+	/// Returns the last continuous data request
+	AMDSClientRelativeCountPlusCountDataRequest* lastContinuousDataRequest() const;
+	/// Sets the continuous data window for the next AMDSClientDataRequest
+	bool setContinuousDataWindow(double continuousDataWindowSeconds);
+	/// Returns the underlying polling rate for the AMDS implementation, usually 1ms
+	int amdsPollingBaseTimeMilliseconds() const;
+
+	/// Returns whether the scaler is currently scanning.
+	bool isStarted() const;
+	/// Returns whether the scaler is set to be in continuous mode or single shot mode.
+	double dwellMode() const;
+	/// Return the current dwell time.  Returns the value in seconds.
+	double dwellTime() const;
+	/// Returns the dwell time tolerance.  Passes the tolerance from the dwell time control.
+	double dwellTimeTolerance() const;
+	/// Returns the current acquisition state
+	double dwellState() const;
+
+	/// Returns the number of channels that are currently enabled
+	int enabledChannelCount() const;
+
+	/// Returns the individual scaler channel provided by \param index.
+	CLSAMDSScalerChannel* channelAt(int index);
```

Comment:
Okay. This has always been a point of confusion for me. Since we're returning a pointer to member owned by the class, if we don't return a const pointer then the class will call the function const but the programmer could call non-const functions on the pointer ... effectively taking non-const action against the class as a whole. I'm more than happy to change the function declarations, I just want to make sure I'm on the right page for when to and when not to.